File: ui/src/components/mascot-pose.ts
// Pure pose model shared by the mascot animator and canvas renderer.

export type MascotMood =
  | "idle"
  | "curious"
  | "thinking"
  | "working"
  | "happy"
  | "celebrating"
  | "sad"
  | "sleepy"
  | "attentive";

type MascotEffect = "none" | "sparkles" | "zzz" | "sparks" | "sweat";

export type MascotPose = {
  floatOffset: number;
  antennaDegrees: number;
  antennaDroop: number;
  leftClawDegrees: number;
  rightClawDegrees: number;
  eyeGlowOpacity: number;
  glowScale: number;
  leftEyeOpenness: number;
  rightEyeOpenness: number;
  happyEyes: number;
  gaze: { x: number; y: number };
  mouthCurve: number;
  mouthOpen: number;
  mouthRound: number;
  blush: number;
  hardHat: number;
  bodyTilt: number;
  bodyStretch: number;
  dizzy: number;
  dizzyPhase: number;
  effect: MascotEffect;
  effectPhase: number;
};

export type MascotPalette = {
  gradientTop: string;
  gradientBottom: string;
  antenna: string;
};

const DARK_PALETTE: MascotPalette = {
  gradientTop: "#ff4d4d",
  gradientBottom: "#991b1b",
  antenna: "#ff4d4d",
};

const LIGHT_PALETTE: MascotPalette = {
  gradientTop: "#ff7079",
  gradientBottom: "#ea4c59",
  antenna: "#ef4b58",
};

export function createMascotPose(): MascotPose {
  return {
    floatOffset: 0,
    antennaDegrees: 0,
    antennaDroop: 0,
    leftClawDegrees: 0,
    rightClawDegrees: 0,
    eyeGlowOpacity: 1,
    glowScale: 1,
    leftEyeOpenness: 1,
    rightEyeOpenness: 1,
    happyEyes: 0,
    gaze: { x: 0, y: 0 },
    mouthCurve: 0,
    mouthOpen: 0,
    mouthRound: 0,
    blush: 0,
    hardHat: 0,
    bodyTilt: 0,
    bodyStretch: 1,
    dizzy: 0,
    dizzyPhase: 0,
    effect: "none",
    effectPhase: 0,
  };
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

/** Keep every channel inside the drawable 120x120 art-space bounds. */
export function clampMascotPose(pose: MascotPose): MascotPose {
  pose.floatOffset = clamp(pose.floatOffset, -12, 2);
  pose.antennaDegrees = clamp(pose.antennaDegrees, -14, 14);
  pose.antennaDroop = clamp(pose.antennaDroop, 0, 1);
  pose.leftClawDegrees = clamp(pose.leftClawDegrees, -45, 45);
  pose.rightClawDegrees = clamp(pose.rightClawDegrees, -45, 45);
  pose.eyeGlowOpacity = clamp(pose.eyeGlowOpacity, 0, 1);
  pose.glowScale = clamp(pose.glowScale, 0.5, 1.6);
  pose.leftEyeOpenness = clamp(pose.leftEyeOpenness, 0, 1);
  pose.rightEyeOpenness = clamp(pose.rightEyeOpenness, 0, 1);
  pose.happyEyes = clamp(pose.happyEyes, 0, 1);
  pose.gaze.x = clamp(pose.gaze.x, -1.2, 1.2);
  pose.gaze.y = clamp(pose.gaze.y, -1.2, 1.2);
  pose.mouthCurve = clamp(pose.mouthCurve, -1, 1);
  pose.mouthOpen = clamp(pose.mouthOpen, 0, 1);
  pose.mouthRound = clamp(pose.mouthRound, 0, 1);
  pose.blush = clamp(pose.blush, 0, 1);
  pose.hardHat = clamp(pose.hardHat, 0, 1);
  pose.bodyTilt = clamp(pose.bodyTilt, -8, 8);
  pose.bodyStretch = clamp(pose.bodyStretch, 0.86, 1.05);
  pose.dizzy = clamp(pose.dizzy, 0, 1);
  return pose;
}

/** Motionless mood expression used when reduced motion is requested. */
export function staticMascotPose(mood: MascotMood): MascotPose {
  const pose = createMascotPose();
  switch (mood) {
    case "idle":
    case "curious":
    case "attentive":
      break;
    case "thinking":
      pose.gaze = { x: 0.3, y: -0.5 };
      break;
    case "working":
      pose.hardHat = 1;
      pose.rightClawDegrees = -28;
      pose.gaze = { x: 0.4, y: 0.35 };
      pose.mouthCurve = 0.15;
      pose.bodyTilt = 2;
      break;
    case "happy":
      pose.mouthCurve = 0.6;
      pose.happyEyes = 0.4;
      break;
    case "celebrating":
      pose.mouthCurve = 0.9;
      pose.mouthOpen = 0.4;
      pose.happyEyes = 0.8;
      pose.leftClawDegrees = 30;
      pose.rightClawDegrees = -30;
      break;
    case "sad":
      pose.antennaDroop = 0.75;
      pose.mouthCurve = -0.55;
      pose.eyeGlowOpacity = 0.6;
      pose.gaze = { x: 0, y: 0.5 };
      break;
    case "sleepy":
      pose.leftEyeOpenness = 0.25;
      pose.rightEyeOpenness = 0.25;
      pose.eyeGlowOpacity = 0.5;
      pose.antennaDroop = 0.35;
      break;
  }
  return pose;
}

export function mascotPalette(light: boolean): MascotPalette {
  return light ? LIGHT_PALETTE : DARK_PALETTE;
}

File: apps/android/app/src/main/java/ai/openclaw/app/ui/design/MascotPose.kt
package ai.openclaw.app.ui.design

enum class MascotMood {
  Idle,
  Curious,
  Thinking,
  Working,
  Happy,
  Celebrating,
  Sad,
  Sleepy,
  Attentive,
}

enum class MascotEffect {
  None,
  Sparkles,
  Zzz,
  Sparks,
  Sweat,
}

data class MascotGaze(
  val x: Double = 0.0,
  val y: Double = 0.0,
)

data class MascotPose(
  var floatOffset: Double = 0.0,
  var antennaDegrees: Double = 0.0,
  var antennaDroop: Double = 0.0,
  var leftClawDegrees: Double = 0.0,
  var rightClawDegrees: Double = 0.0,
  var eyeGlowAlpha: Double = 1.0,
  var glowScale: Double = 1.0,
  var leftEyeOpenness: Double = 1.0,
  var rightEyeOpenness: Double = 1.0,
  var happyEyes: Double = 0.0,
  var gaze: MascotGaze = MascotGaze(),
  var mouthCurve: Double = 0.0,
  var mouthOpen: Double = 0.0,
  var mouthRound: Double = 0.0,
  var blush: Double = 0.0,
  var hardHat: Double = 0.0,
  var bodyTilt: Double = 0.0,
  var bodyStretch: Double = 1.0,
  var effect: MascotEffect = MascotEffect.None,
  var effectPhase: Double = 0.0,
) {
  /** Keeps every channel inside the drawable 120x120 art-space bounds. */
  fun clamp(): MascotPose {
    floatOffset = floatOffset.coerceIn(-12.0, 2.0)
    antennaDegrees = antennaDegrees.coerceIn(-14.0, 14.0)
    antennaDroop = antennaDroop.coerceIn(0.0, 1.0)
    leftClawDegrees = leftClawDegrees.coerceIn(-45.0, 45.0)
    rightClawDegrees = rightClawDegrees.coerceIn(-45.0, 45.0)
    eyeGlowAlpha = eyeGlowAlpha.coerceIn(0.0, 1.0)
    glowScale = glowScale.coerceIn(0.5, 1.6)
    leftEyeOpenness = leftEyeOpenness.coerceIn(0.0, 1.0)
    rightEyeOpenness = rightEyeOpenness.coerceIn(0.0, 1.0)
    happyEyes = happyEyes.coerceIn(0.0, 1.0)
    gaze = MascotGaze(gaze.x.coerceIn(-1.2, 1.2), gaze.y.coerceIn(-1.2, 1.2))
    mouthCurve = mouthCurve.coerceIn(-1.0, 1.0)
    mouthOpen = mouthOpen.coerceIn(0.0, 1.0)
    mouthRound = mouthRound.coerceIn(0.0, 1.0)
    blush = blush.coerceIn(0.0, 1.0)
    hardHat = hardHat.coerceIn(0.0, 1.0)
    bodyTilt = bodyTilt.coerceIn(-8.0, 8.0)
    bodyStretch = bodyStretch.coerceIn(0.86, 1.05)
    return this
  }

  companion object {
    /** Motionless mood expression used when Android animations are disabled. */
    fun staticPose(mood: MascotMood): MascotPose =
      MascotPose().apply {
        when (mood) {
          MascotMood.Idle,
          MascotMood.Curious,
          MascotMood.Attentive,
          -> Unit
          MascotMood.Thinking -> gaze = MascotGaze(x = 0.3, y = -0.5)
          MascotMood.Working -> {
            hardHat = 1.0
            rightClawDegrees = -28.0
            gaze = MascotGaze(x = 0.4, y = 0.35)
            mouthCurve = 0.15
            bodyTilt = 2.0
          }
          MascotMood.Happy -> {
            mouthCurve = 0.6
            happyEyes = 0.4
          }
          MascotMood.Celebrating -> {
            mouthCurve = 0.9
            mouthOpen = 0.4
            happyEyes = 0.8
            leftClawDegrees = 30.0
            rightClawDegrees = -30.0
          }
          MascotMood.Sad -> {
            antennaDroop = 0.75
            mouthCurve = -0.55
            eyeGlowAlpha = 0.6
            gaze = MascotGaze(x = 0.0, y = 0.5)
          }
          MascotMood.Sleepy -> {
            leftEyeOpenness = 0.25
            rightEyeOpenness = 0.25
            eyeGlowAlpha = 0.5
            antennaDroop = 0.35
          }
        }
      }
  }
}

fun staticPose(mood: MascotMood): MascotPose = MascotPose.staticPose(mood)

/**
 * Tinted silhouettes stay on the ambient idle loop: mood faces cannot read in
 * monochrome, and tiny tinted toolbar marks must not hammer or celebrate.
 */
internal fun effectiveMascotMood(
  mood: MascotMood,
  tinted: Boolean,
): MascotMood = if (tinted) MascotMood.Idle else mood
